File: src/util_paths.py
"""Resolve personal-asset paths with private-dir-first lookup.

Each Stand has its own identity + avatar. These files are gitignored and
machine-local. Canonical home is `$SUTANDO_MEMORY_DIR/machine-<hostname>/`
so they live with the rest of the per-machine memory under the private
sync repo. Public-workspace fallback is preserved so existing installs
keep working until they migrate.

The env var `SUTANDO_MEMORY_DIR` is the canonical name per the 2026-05-18
workspace-design RFC (#858, Decision 2). The legacy name `SUTANDO_PRIVATE_DIR`
is honored as a fallback for one release with a deprecation warning on
every read (cron environments miss startup-only warnings, so logging at
every resolution is intentional).

Usage:
    from util_paths import personal_path
    si = personal_path("stand-identity.json")
    avatar = personal_path("stand-avatar.png")  # also tries assets/ in public
"""
from __future__ import annotations
import os
import re
import socket
import subprocess
import sys
from pathlib import Path

def _memory_dir_env() -> str | None:
    """Return the resolved memory-dir env value, preferring the new name.

    Lookup order:
      1. `SUTANDO_MEMORY_DIR` (canonical post-#858 / #870)
      2. `SUTANDO_PRIVATE_DIR` (legacy, with deprecation warning emitted
         to stderr on every read — not just once at startup; cron and
         launchd environments miss startup-only warnings).

    Returns the raw env value (caller must `os.path.expanduser` if needed),
    or None when neither is set."""
    new = os.environ.get("SUTANDO_MEMORY_DIR")
    if new:
        return new
    legacy = os.environ.get("SUTANDO_PRIVATE_DIR")
    if legacy:
        # Every-read deprecation warning. This is loud by design — the
        # legacy alias will drop in the next release and silent users
        # would otherwise miss the cutover. See #870 for the rename plan.
        print(
            "[util_paths.py] DEPRECATION: SUTANDO_PRIVATE_DIR is the old name "
            "for the memory dir; set SUTANDO_MEMORY_DIR instead (this alias "
            "will be removed in the next release). See #870.",
            file=sys.stderr,
        )
        return legacy
    return None


def _workspace_root() -> Path:
    """Workspace root for runtime-state paths.

    Delegates to workspace_default.resolve_workspace() so the post-v0.8
    canonical default (<repo>/workspace/) is honored. ($SUTANDO_WORKSPACE is
    no longer honored for resolution per #1440; see `src/sutando_config.py`.)

    `migrate=False` — path resolution shouldn't trigger migrations on
    every call. Migration runs from src/startup.sh and the bridge boot
    paths where it belongs.
    """
    try:
        from workspace_default import resolve_workspace
        return resolve_workspace(migrate=False)
    except ImportError:
        # workspace_default not on sys.path (standalone invocation outside the
        # repo). Locate the repo root via git, then call sutando-config.sh to
        # honor sutando.config.local.json and the M0 override chain.
        try:
            toplevel = subprocess.run(
                ["git", "rev-parse", "--show-toplevel"],
                capture_output=True, text=True, timeout=5,
            )
            if toplevel.returncode == 0 and toplevel.stdout.strip():
                _sh = Path(toplevel.stdout.strip()) / "scripts" / "sutando-config.sh"
                result = subprocess.run(
                    ["bash", str(_sh), "workspace"],
                    capture_output=True, text=True, timeout=5,
                )
                if result.returncode == 0 and result.stdout.strip():
                    return Path(result.stdout.strip())
        except Exception:
            pass
        # Last-ditch default: the canonical post-v0.8 home-dir location
        # (~/sutando-workspace, matching sutando_config.py resolve_workspace) —
        # NOT the pre-v0.8 dotted ~/.sutando/workspace, which no longer exists.
        return Path.home() / "sutando-workspace"


def _host_label() -> str:
    r"""Per-host directory label. Precedence:
      1. `$SUTANDO_HOST_LABEL` (or legacy `$SUTANDO_HOST_OVERRIDE`)
      2. macOS `scutil --get LocalHostName` (the stable Bonjour name)
      3. short `hostname`

    Why scutil before hostname: on DHCP networks `hostname` can drift (e.g. a
    Comcast residential lease returns `Chis-MBP.hsd1.wa.comcast.net` →
    `Chis-MBP`), while the Bonjour LocalHostName (`Chis-MacBook-Pro`) is stable.
    A drifting `hostname` splits per-host paths — two `hosts/<label>/` dirs,
    phantom `state/cores/<label>.alive`, and `personal_path()` falling back to
    the workspace root (2026-06-22 incident). `scutil` is macOS-only; on Linux
    it's absent and we fall through to `hostname`.

    Single source of truth for the per-host segment so the legacy
    `machine-<host>/` (memory-dir) and new `hosts/<host>/` (workspace)
    conventions stay in lockstep. Kept in lockstep with `_host()` in
    sync-workspace.sh (same precedence)."""
    env = os.environ.get("SUTANDO_HOST_LABEL") or os.environ.get("SUTANDO_HOST_OVERRIDE")
    # Strip before testing: a blank-but-set override (`SUTANDO_HOST_LABEL=" "`,
    # trivially produced by an unquoted expansion in a launcher) is truthy in
    # Python, so `if env:` returned the whitespace itself as the label. That
    # yields `hosts/   /` and `state/cores/   .alive` — the same per-host path
    # split as the 2026-06-22 DHCP-drift incident above, but self-inflicted and
    # far harder to spot in a directory listing. Blank means "not set": fall
    # through to scutil/hostname. Matches SutandoConfig.hostLabel() in
    # src/Sutando/main.swift, which already trims.
    env = (env or "").strip()
    if env:
        return env
    try:
        out = subprocess.run(
            ["scutil", "--get", "LocalHostName"],
            capture_output=True, text=True, timeout=2,
        )
        if out.returncode == 0 and out.stdout.strip():
            return out.stdout.strip()
    except (OSError, subprocess.SubprocessError):
        pass
    return socket.gethostname().split(".")[0]


def _private_machine_dir() -> Path | None:
    root = _memory_dir_env()
    if not root:
        return None
    expanded = os.path.expanduser(root)
    return Path(expanded) / f"machine-{_host_label()}"


def personal_path(filename: str, workspace: Path | None = None) -> Path:
    """Resolve a personal-asset path.

    Order: `<workspace>/hosts/<host>/<filename>` (new per-host home, #1717)
    → `$SUTANDO_MEMORY_DIR/machine-<host>/<filename>` (legacy memory-dir
    per-host) → `<workspace>/<filename>`.
    (Legacy `$SUTANDO_PRIVATE_DIR` is honored as a fallback with a
    deprecation warning — see `_memory_dir_env()`.)
    For files known to live under `assets/` in the public workspace
    (currently `stand-avatar.png`), also tries `<workspace>/assets/<filename>`
    before falling back to `<workspace>/<filename>`.

    The `hosts/<host>/` probe is read-side only and purely additive: when no
    such file exists, resolution is identical to the pre-#1717 behavior. This
    is the reader half of the per-host relocation — without it, moving a
    per-host file into `hosts/<host>/` would silently strand readers on the
    workspace-root fallback (the H4 regression).

    Returns the FIRST existing path. If none exist, returns the preferred
    private-dir path so the caller's `.exists()` check fails gracefully.
    """
    explicit_ws = workspace is not None
    ws = workspace if explicit_ws else _workspace_root()

    # New per-host canonical home (workspace-as-git-repo, #1717). Probed first
    # so relocated files are found; absent → falls through to legacy order.
    host_dir = ws / "hosts" / _host_label()
    p = host_dir / filename
    if p.exists():
        return p

    private = _private_machine_dir()
    if private is not None:
        p = private / filename
        if p.exists():
            return p

    # Public workspace — assets/ first for avatar-style files, then root
    if filename in {"stand-avatar.png"}:
        p = ws / "assets" / filename
        if p.exists():
            return p

    p = ws / filename
    if p.exists():
        return p

    # Nothing exists — this is the path the caller will CREATE. It must stay
    # inside the workspace the caller named, or their isolation is a fiction.
    #
    # The read probes above may legitimately return a legacy machine-<host>/
    # file: that is the migration fallback and it is load-bearing while those
    # files still exist. But the WRITE destination is a different question, and
    # answering it from $SUTANDO_MEMORY_DIR ignored the argument entirely — so
    # `personal_path("x.md", <fresh tmp>)` handed back a path in the operator's
    # real, vault-SYNCED memory tree. A test that passed a tmpdir believing it
    # was isolated wrote into Chi's vault instead; ALPHA/BRAVO fixtures reached
    # it that way (#2452). Passing a tmpdir is not isolation unless the resolved
    # path is actually inside it.
    #
    # When no workspace was passed the caller has accepted ambient resolution,
    # so the env-based private dir remains correct and behavior is unchanged.
    # The escape exists ONLY when a private dir is configured: that is the branch
    # that answered from the environment instead of from `ws`. When it is None the
    # code already fell through to `ws / filename`, which is inside the workspace
    # and needs no change — and `tests/util-paths-hosts-resolution.test.py` pins
    # that as a deliberate #1717 decision ("the fix is read-side only; write
    # target is untouched"). So keep the write target where #1717 put it, the
    # workspace root, and change only WHOSE workspace answers.
    if private is not None:
        if explicit_ws:
            return ws / filename
        return private / filename
    if filename in {"stand-avatar.png"}:
        return ws / "assets" / filename
    return ws / filename


def shared_personal_path(filename: str, workspace: Path | None = None) -> Path:
    """Resolve a shared-private path (notes, build_log, etc.) — files that
    sync across all of an owner's machines, not per-machine state.

    Order: `$SUTANDO_MEMORY_DIR/<filename>` (top-level, shared) → `<workspace>/<filename>`.
    (Legacy `$SUTANDO_PRIVATE_DIR` is honored as a fallback with a
    deprecation warning — see `_memory_dir_env()`.)

    Difference vs `personal_path`: this resolves to the top-level private dir,
    NOT `machine-<host>/`. Use for files like notes/, where every Mac in
    Chi's fleet should see the same content.

    Returns the FIRST existing path. If none exist, returns the preferred
    private path so the caller's `.exists()` check fails gracefully.
    """
    ws = workspace if workspace is not None else _workspace_root()

    root = _memory_dir_env()
    if root:
        private = Path(os.path.expanduser(root)) / filename
        if private.exists():
            return private
        # Fall back to workspace if private doesn't have it, but remember
        # the preferred private path for the "nothing exists" branch.
        p = ws / filename
        if p.exists():
            return p
        return private

    p = ws / filename
    return p


# ---------------------------------------------------------------------------
# Claude Code home directory — the host CLI's per-user state lives at
# `~/.claude/`. Sutando consumes several subpaths (channels/, projects/,
# skills/, settings.json, etc.); centralizing the resolution here keeps the
# host-CLI dependency surface a single grep.
#
# Why this helper: per the 2026-05-18 workspace-design RFC discussion, the
# dependency on `~/.claude/` is real (memory storage, channel tokens, skill
# discovery, slash-command write convention) and we accept it operationally —
# but we want the surface countable so a future swap is a 1-day grep+replace
# rather than a re-architecture. ANY new read/write into the Claude Code home
# directory should go through this helper.
#
# Resolution: prefer $CLAUDE_HOME if set (override / testing), else
# `~/.claude/`. Does NOT create the dir.
# ---------------------------------------------------------------------------

def claude_home_path(*subpath: str, vanilla: bool = False) -> Path:
    """Resolve a path under Claude Code's per-user home (`~/.claude/` by default).

    Pass subpath components positionally, e.g.:
        claude_home_path("channels", "discord", "access.json")
        claude_home_path("projects", project_slug, "memory", "MEMORY.md")
        claude_home_path("skills", skill_name)

    Resolution order:
      1. $CLAUDE_CONFIG_DIR (M2 workspace-scoped path; set by the
         `claude-sutando` shell function + start-cli.sh — when present,
         bridges + memory readers see the workspace's .claude-sutando/).
      2. $CLAUDE_HOME (legacy alt-host override, kept for tests).
      3. ~/.claude/ (default — vanilla `claude` users).

    `vanilla=True` asks a DIFFERENT question: where does *stock* Claude Code
    keep this, regardless of where Sutando relocated its own config? It skips
    $CLAUDE_CONFIG_DIR entirely and reads $SOURCE_CLAUDE_CONFIG_DIR (the name
    the note below already gives that location) before the default. Use it ONLY
    to read state written by a tool that hardcodes the stock home and does not
    honour $CLAUDE_CONFIG_DIR — that surface stays as countable as this one.
    A caller who wants the relocated dir wants the default form; collapsing the
    two is the bug sonichi#2629 is about, where the official discord plugin
    wrote approval markers to the stock home and the bridge read the relocated
    one, so confirmations were silently never sent.

    The CLAUDE_CONFIG_DIR check goes first because for a claude-sutando
    install, that's where settings, sessions, channels, skills, and memory
    actually live post-migrate. The CLAUDE_HOME hatch still works for tests
    that need a non-default but non-workspace location.

    Companion env var: $SOURCE_CLAUDE_CONFIG_DIR (defaults to ~/.claude) is
    used by migration scripts (sutando-shell-setup.sh --migrate, src/migrate.sh)
    to refer to the READ-FROM source — i.e., where vanilla claude state lives
    historically. claude_home_path() does NOT consult it; this helper is for
    RUNTIME path resolution. Migration code uses SOURCE_CLAUDE_CONFIG_DIR
    directly to keep the read-side / write-side distinction visible.
    """
    ccd_env = None if vanilla else os.environ.get("CLAUDE_CONFIG_DIR")
    home_env = (os.environ.get("SOURCE_CLAUDE_CONFIG_DIR") if vanilla
                else os.environ.get("CLAUDE_HOME"))
    if ccd_env:
        base = Path(os.path.expanduser(ccd_env))
    elif home_env:
        base = Path(os.path.expanduser(home_env))
    else:
        # `vanilla=True` landing here is the CORRECT answer, not a fallback, so
        # it must not fire the "nothing is configured" banner.
        if not vanilla:
            _emit_claude_home_fallback_banner_once()
        base = Path.home() / ".claude"
    if not subpath:
        return base
    return base.joinpath(*subpath)


# ---------------------------------------------------------------------------
# Claude Code project slug
#
# Claude Code stores per-project state at `<claude-home>/projects/<slug>/`,
# where `<slug>` is a slugified form of the project's filesystem path. The
# slugification rule is Claude Code's own private detail — we consume it, we
# do not define it. Resolve a slug with project_slug(); never re-derive one
# inline.
# ---------------------------------------------------------------------------

def slug_derivation_key(name: str) -> str:
    """Collapse a Claude project slug to a derivation-INDEPENDENT key.

    Two slugs describing the SAME path agree once every run of
    non-alphanumerics is collapsed to one ``-`` and case is folded, while an
    unrelated project does not collide::

        -Users-me-Library-Application-Support-space-ag2-app-engine-sutando
        -Users-me-Library-Application Support-space.ag2.app-engine-sutando
            -> users-me-library-application-support-space-ag2-app-engine-sutando  (same)
        -Users-me-Documents-unrelated-repo
            -> users-me-documents-unrelated-repo                                  (different)
    """
    return re.sub(r"[^a-z0-9]+", "-", name.lower()).strip("-")


def _slug_formula(path: "str | Path") -> str:
    """Best-effort reproduction of Claude Code's slugification.

    Only used when discovery finds nothing (see project_slug). Maps ``/``,
    ``.`` and spaces to ``-``, which is what the current Claude Code produces
    for the bundled-install path this fix came from::

        /Users/me/Library/Application Support/space.ag2.app/engine/sutando
            -> -Users-me-Library-Application-Support-space-ag2-app-engine-sutando
    """
    return re.sub(r"[/. ]", "-", str(path))


def project_slug(path: "str | Path", *, vanilla: bool = False) -> str:
    """Resolve the Claude Code project-dir slug for a filesystem path.

    **Discovery first, formula second — deliberately.**

    Every previous version of this logic hardcoded a formula, and each one
    was a prediction about a detail Claude Code owns and has already changed.
    Both of these are real project dirs on one machine, for the same
    ``/var/folders`` family::

        -var-folders-bs-0nw4fhvs4599_zcgpk3fbcnh0000gn-T-agent-state-...   (`_` kept)
        -private-var-folders-bs-0nw4fhvs4599-zcgpk3fbcnh0000gn-T          (`_` mapped)

    A formula that is right today silently stops being right after a Claude
    Code upgrade — and it fails in the worst possible way, because nothing
    errors: the wrong slug's directory is simply *created* on first write, so
    memory splits across parallel stores and every path-derived reader looks
    at whichever one it happened to compute. That is precisely how one repo
    came to own three populated project dirs (sonichi#2723).

    So: look for a project dir that ALREADY EXISTS and describes the same
    path, comparing through slug_derivation_key() so the comparison does not
    itself depend on a formula. Use that real directory when found. Fall back
    to _slug_formula() only when Claude Code has not created one yet — the
    first-run case, where any answer is a guess and creating the directory is
    the point.

    On ties (more than one existing dir matching the key — the split this
    function exists to stop recurring) prefer the most recently modified,
    which is the one Claude Code is actually writing to.
    """
    target = slug_derivation_key(str(path))
    try:
        projects = claude_home_path("projects", vanilla=vanilla)
        matches = [
            entry for entry in projects.iterdir()
            if entry.is_dir() and slug_derivation_key(entry.name) == target
        ]
        if matches:
            return max(matches, key=lambda p: p.stat().st_mtime).name
    except OSError:
        # No projects dir yet, or unreadable — fall through to the formula.
        pass
    return _slug_formula(path)


def channel_access_path(source: str) -> Path:
    """Resolve `channels/<source>/access.json` with the ~30-day legacy fallback.

    Prefer the canonical claude_home_path() location. If that file does NOT
    exist but the pre-migration `~/.claude/channels/<source>/access.json`
    does, return the legacy path and emit a one-line stderr deprecation
    warning — per the CLAUDE.md migration policy (readers prefer canonical,
    fall back to legacy for ~30 days).

    Why this exists: bridges restarted under a fresh $CLAUDE_CONFIG_DIR
    before the channel-bridge migrate step copies channels/ would otherwise
    see no access.json at all — Telegram/Slack then re-arm TOFU onboarding
    and the next DM sender auto-enrolls as owner. Falling back to the
    populated legacy allowlist keeps access control continuous across the
    migration window. Writers (TOFU onboarding, /discord:access) use the
    same resolved path, so the legacy file stays the single source of truth
    until it is actually migrated.
    """
    canonical = claude_home_path("channels", source, "access.json")
    if canonical.exists():
        return canonical
    legacy = Path.home() / ".claude" / "channels" / source / "access.json"
    if legacy != canonical and legacy.exists():
        print(
            f"[util_paths] DEPRECATION: using legacy {legacy} — canonical "
            f"{canonical} missing. Run the channel-bridge migrate step "
            f"(scripts/sutando-migrate.sh) to relocate; this fallback is "
            f"removed ~30 days post-migration.",
            file=sys.stderr,
        )
        return legacy
    return canonical


# ---------------------------------------------------------------------------
# Fallback-banner gate — fires ONCE per process when claude_home_path() lands
# on the ~/.claude/ default because neither $CLAUDE_CONFIG_DIR nor $CLAUDE_HOME
# was set. Owner directive #design 2026-06-07 (Option A+ for channels migration):
# the silent ~/.claude/ fallback was load-bearing for any boot path that forgot
# to set CCD; the banner makes that miswiring visible without forcing a hard
# error in the deprecation window. Banner is suppressible via
# $SUTANDO_SUPPRESS_CCD_FALLBACK_BANNER=1 for tests / scripts that intentionally
# exercise the ~/.claude/ path.
# ---------------------------------------------------------------------------

_CLAUDE_HOME_FALLBACK_BANNER_FIRED = False


def _emit_claude_home_fallback_banner_once() -> None:
    global _CLAUDE_HOME_FALLBACK_BANNER_FIRED
    if _CLAUDE_HOME_FALLBACK_BANNER_FIRED:
        return
    if os.environ.get("SUTANDO_SUPPRESS_CCD_FALLBACK_BANNER") == "1":
        _CLAUDE_HOME_FALLBACK_BANNER_FIRED = True
        return
    _CLAUDE_HOME_FALLBACK_BANNER_FIRED = True
    print(
        "claude_home_path: $CLAUDE_CONFIG_DIR not set — falling back to ~/.claude/. "
        "Set CLAUDE_CONFIG_DIR before starting Sutando services (the `claude-sutando` "
        "shell function and src/startup.sh set it; ad-hoc launches must too) so "
        "channels/skills/hooks/sessions resolve to the workspace-scoped per-runtime "
        "location post-#1454. Suppress with SUTANDO_SUPPRESS_CCD_FALLBACK_BANNER=1.",
        file=sys.stderr,
    )

def write_private_text(path: "Path", text: str) -> None:
    """Write ``text`` to ``path`` as an owner-only (0600) file, with no window.

    ``Path.write_text()`` creates at the process umask — commonly 0644 — so the
    familiar ``write_text(...)`` + ``os.chmod(..., 0o600)`` pair leaves the file
    world-readable for the interval between the two calls. For access-control
    data (allowlists, owner ids, tier maps) that interval is the whole exposure.

    ``os.open`` applies the mode at CREATION, and ``fchmod`` covers the case
    where the file already existed (the mode argument is ignored then). O_TRUNC
    rather than O_EXCL deliberately: this is used on best-effort paths wrapped in
    broad excepts, and O_EXCL would turn a leftover temp file from a crash into a
    permanent silent failure.
    """
    # Order matters for DATA INTEGRITY, not just permissions. O_TRUNC empties the
    # file at OPEN, so with `O_CREAT|O_TRUNC` then fchmod, a hardening failure
    # leaves an EXISTING backup empty -- a permission error would destroy exactly
    # the durable copy that exists to survive a wipe. (The old
    # write_text()+chmod at least failed with correct data and loose perms.)
    #
    # So: open WITHOUT O_TRUNC, harden first, and only truncate once nothing can
    # still fail destructively.
    fd = os.open(path, os.O_WRONLY | os.O_CREAT, 0o600)
    try:
        os.fchmod(fd, 0o600)   # existing file: mode arg above was ignored
        os.ftruncate(fd, 0)    # nothing destroyed until hardening succeeded
    except BaseException:
        os.close(fd)
        raise
    with os.fdopen(fd, "w") as fh:  # fdopen takes ownership of fd from here
        fh.write(text)

File: src/util_paths.ts
// TypeScript twin of src/util_paths.py — personal-asset path resolution.
//
// Two helpers, one for per-machine state, one for shared-across-fleet state:
//
//   personalPath(filename)          — `$SUTANDO_MEMORY_DIR/machine-<host>/<filename>`
//                                     For files where each Mac has its own copy
//                                     (stand-identity.json, pending-questions.md).
//
//   sharedPersonalPath(filename)    — `$SUTANDO_MEMORY_DIR/<filename>`
//                                     For files synced across the whole fleet
//                                     (notes/, build_log.md).
//
// Both fall back to `<workspace>/<filename>` so existing installs keep working
// until they migrate. The `workspace` arg is optional; when omitted, the
// helpers resolve via `resolveWorkspace()` — post-v0.8 (#1440) the default is
// `<repo>/workspace/` and `$SUTANDO_WORKSPACE` is no longer honored — NOT
// process.cwd(). Pre-#839 fixes the fallback was cwd, which silently produced
// the wrong path on hosts where the caller's cwd drifted from the workspace dir.
//
// Env var `SUTANDO_MEMORY_DIR` is the canonical name post-#858 / #870. The
// legacy alias `SUTANDO_PRIVATE_DIR` is honored as a fallback for one release
// with a deprecation warning logged to stderr on every read (cron / launchd
// environments miss startup-only warnings, so logging at every resolution is
// intentional).

import { execFileSync } from 'node:child_process';
import { existsSync, readFileSync, readdirSync, statSync } from 'node:fs';
import { hostname } from 'node:os';
import { join } from 'node:path';
import { resolveWorkspace } from './workspace_default.js';

function expandHome(p: string): string {
	return p.replace(/^~/, process.env.HOME || '');
}

/**
 * Return the resolved memory-dir env value, preferring the new name.
 *
 * Lookup order:
 *   1. `SUTANDO_MEMORY_DIR` (canonical post-#858 / #870)
 *   2. `SUTANDO_PRIVATE_DIR` (legacy, with deprecation warning emitted to
 *      stderr on every read — not just once at startup; cron and launchd
 *      environments miss startup-only warnings).
 *
 * Returns the raw env value (caller must expandHome if needed), or undefined
 * when neither is set.
 */
export function memoryDirEnv(): string | undefined {
	const next = process.env.SUTANDO_MEMORY_DIR;
	if (next) return next;
	const legacy = process.env.SUTANDO_PRIVATE_DIR;
	if (legacy) {
		// Every-read deprecation warning. Loud by design — the legacy alias
		// will drop in the next release and silent users would otherwise miss
		// the cutover. See #870 for the rename plan.
		console.warn(
			'[util_paths.ts] DEPRECATION: SUTANDO_PRIVATE_DIR is the old name ' +
				'for the memory dir; set SUTANDO_MEMORY_DIR instead (this alias ' +
				'will be removed in the next release). See #870.',
		);
		return legacy;
	}
	return undefined;
}

/**
 * Read macOS Bonjour `LocalHostName` via scutil. Returns '' when unavailable
 * (non-macOS, scutil missing/errored, or an empty result) so the caller falls
 * through to the next precedence step. Never throws.
 */
function scutilLocalHostName(): string {
	try {
		return execFileSync('scutil', ['--get', 'LocalHostName'], {
			timeout: 2000,
			stdio: ['ignore', 'pipe', 'ignore'],
		})
			.toString()
			.trim();
	} catch {
		return '';
	}
}

/**
 * Per-host directory label. Precedence (mirrors `_host_label()` in
 * util_paths.py and `_host()` in sync-workspace.sh — the single source of
 * truth for the per-host segment):
 *   1. `$SUTANDO_HOST_LABEL` (or legacy `$SUTANDO_HOST_OVERRIDE`), trimmed;
 *      blank-after-trim counts as unset.
 *   2. macOS `scutil --get LocalHostName` — the STABLE Bonjour name.
 *   3. short `hostname` (mDNS/domain suffix stripped) — last resort.
 *
 * Step 2 is load-bearing: on DHCP networks `hostname` can drift (e.g. an
 * AT&T/Comcast lease → `QingyunsMBP2200` / `Chis-MBP`) while the Bonjour
 * `LocalHostName` (`Qingyuns-MacBook-Pro-2200` / `Chis-MacBook-Pro`) is stable.
 * Without it, TS-side per-host resolution diverges from the py/bash side —
 * two `hosts/<label>/` dirs, and `personalPath()` reading a ghost dir the
 * writers never populate (silent per-host data loss, #1745). An explicit
 * label is used RAW (a dotted label like "a.b" must NOT be split — that would
 * strand the reader); only the auto-detected hostname has its suffix stripped.
 *
 * `scutil`/`rawHostname` are injectable so tests exercise every branch without
 * a real macOS.
 */
export function resolveHostLabel(
	env: NodeJS.ProcessEnv = process.env,
	scutil: () => string = scutilLocalHostName,
	rawHostname: string = hostname(),
): string {
	// Trim before testing: a blank-but-set override is TRUTHY in JS, so
	// `if (label)` returned the whitespace itself as the label — `hosts/   /`,
	// the same self-inflicted per-host split the DHCP note above describes.
	// Blank means "not set": fall through to scutil/hostname. Lockstep with
	// _host_label() in util_paths.py and _host() in sync-workspace.sh.
	const label = (env.SUTANDO_HOST_LABEL || env.SUTANDO_HOST_OVERRIDE || '').trim();
	if (label) return label;
	const bonjour = scutil();
	if (bonjour) return bonjour;
	return rawHostname.split('.')[0];
}

function hostLabel(): string {
	return resolveHostLabel();
}

/** Per-machine resolver. */
export function personalPath(filename: string, workspace?: string): string {
	const ws = workspace ?? resolveWorkspace();
	// New per-host canonical home (workspace-as-git-repo, #1717). Probed first
	// so relocated files are found; absent → falls through to the legacy order
	// (identical to pre-#1717 behavior). Reader half of the per-host
	// relocation — without it, moving a per-host file into `hosts/<host>/`
	// would silently strand readers on the workspace-root fallback (H4).
	const hostCandidate = join(ws, 'hosts', hostLabel(), filename);
	if (existsSync(hostCandidate)) return hostCandidate;
	const privateRoot = memoryDirEnv();
	if (privateRoot) {
		const root = expandHome(privateRoot);
		const candidate = join(root, `machine-${hostLabel()}`, filename);
		if (existsSync(candidate)) return candidate;
	}
	// stand-avatar.png lives under assets/ in the public workspace.
	if (filename === 'stand-avatar.png') {
		const inAssets = join(ws, 'assets', filename);
		if (existsSync(inAssets)) return inAssets;
	}
	const wsPath = join(ws, filename);
	if (existsSync(wsPath)) return wsPath;
	// Nothing exists; return preferred private path so caller's existsSync()
	// check fails gracefully.
	if (privateRoot) {
		const root = expandHome(privateRoot);
		return join(root, `machine-${hostLabel()}`, filename);
	}
	if (filename === 'stand-avatar.png') return join(ws, 'assets', filename);
	return wsPath;
}

/** Shared-across-fleet resolver (top-level private dir, not per-machine). */
export function sharedPersonalPath(filename: string, workspace?: string): string {
	const ws = workspace ?? resolveWorkspace();
	const privateRoot = memoryDirEnv();
	if (privateRoot) {
		const root = expandHome(privateRoot);
		const candidate = join(root, filename);
		if (existsSync(candidate)) return candidate;
		const wsPath = join(ws, filename);
		if (existsSync(wsPath)) return wsPath;
		return candidate;
	}
	return join(ws, filename);
}


// ---------------------------------------------------------------------------
// Claude Code home directory — the host CLI's per-user state lives at
// `~/.claude/`. Sutando consumes several subpaths (channels/, projects/,
// skills/, settings.json, etc.); centralizing the resolution here keeps the
// host-CLI dependency surface a single grep.
//
// Why this helper: per the 2026-05-18 workspace-design RFC discussion, the
// dependency on `~/.claude/` is real (memory storage, channel tokens, skill
// discovery, slash-command write convention) and we accept it operationally —
// but we want the surface countable so a future swap is a 1-day grep+replace
// rather than a re-architecture. ANY new read/write into the Claude Code home
// directory should go through this helper.
//
// Resolution (3-tier, prefer most specific):
//   1. $CLAUDE_CONFIG_DIR  — Claude Code's canonical env var (string present
//      in the `claude` binary). Set by `claude-sutando` shell function +
//      src/agent/claude/cli/start-cli.sh + src/startup.sh so every workspace gets its own
//      .claude-sutando/ tree instead of sharing global ~/.claude/.
//   2. $CLAUDE_HOME        — deprecated legacy override (kept for one release
//      so pre-M0 callers / test fixtures don't break instantly). Emits a
//      one-shot warning to stderr on first read.
//   3. ~/.claude/          — final fallback.
// Does NOT create the dir.
// ---------------------------------------------------------------------------

let _claudeHomeDeprecationWarned = false;

/**
 * Resolve a path under Claude Code's per-user config dir.
 *
 * Pass subpath components as separate args:
 *   claudeHomePath('channels', 'discord', 'access.json')
 *   claudeHomePath('projects', projectSlug, 'memory', 'MEMORY.md')
 *   claudeHomePath('skills', skillName)
 *
 * Prefers `$CLAUDE_CONFIG_DIR` (Claude Code canonical). Falls back to
 * deprecated `$CLAUDE_HOME` then `~/.claude/`. See Mini PR #1415 review #5
 * for the original-env-var-mismatch that motivated this.
 */
export function claudeHomePath(...subpath: string[]): string {
	const ccd = process.env.CLAUDE_CONFIG_DIR;
	const home = process.env.CLAUDE_HOME;
	let base: string;
	if (ccd) {
		base = expandHome(ccd);
	} else if (home) {
		if (!_claudeHomeDeprecationWarned) {
			_claudeHomeDeprecationWarned = true;
			console.warn(
				'[util_paths] $CLAUDE_HOME is deprecated; set $CLAUDE_CONFIG_DIR instead (will be removed next release).',
			);
		}
		base = expandHome(home);
	} else {
		base = join(process.env.HOME || '', '.claude');
	}
	if (subpath.length === 0) return base;
	return join(base, ...subpath);
}

// ---------------------------------------------------------------------------
// Claude Code project slug — TypeScript twin of util_paths.py's project_slug().
//
// Claude Code stores per-project state at `<claude-home>/projects/<slug>/`,
// where `<slug>` is a slugified form of the project path. That slugification
// is Claude Code's own private detail — we consume it, we do not define it.
// Resolve a slug with projectSlug(); never re-derive one inline.
// ---------------------------------------------------------------------------

/**
 * Collapse a Claude project slug to a derivation-INDEPENDENT key.
 *
 * Two slugs describing the SAME path agree once every run of non-alphanumerics
 * is collapsed to one `-` and case is folded; unrelated projects do not collide.
 */
export function slugDerivationKey(name: string): string {
	return name.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-|-$/g, '');
}

/** Best-effort reproduction of Claude Code's slugification (fallback only). */
export function slugFormula(path: string): string {
	return path.replace(/[/. ]/g, '-');
}

/**
 * Resolve the Claude Code project-dir slug for a filesystem path.
 *
 * **Discovery first, formula second — deliberately.** Every previous version
 * of this logic hardcoded a formula, and each was a prediction about a detail
 * Claude Code owns and has already changed (one machine holds both
 * `...0nw4fhvs4599_zcgpk3fbcnh0000gn...` and `...4599-zcgpk3...` for the same
 * `/var/folders` family). A formula that is right today stops being right
 * after an upgrade — and fails silently, because the wrong slug's directory is
 * simply created on first write. See util_paths.py's project_slug() for the
 * full rationale; both twins must stay in step.
 */
export function projectSlug(path: string): string {
	const target = slugDerivationKey(path);
	try {
		const projects = claudeHomePath('projects');
		const matches = readdirSync(projects, { withFileTypes: true })
			.filter(e => e.isDirectory() && slugDerivationKey(e.name) === target)
			.map(e => e.name);
		if (matches.length > 0) {
			// On a split, prefer the most recently modified — the live store.
			return matches.sort(
				(a, b) =>
					statSync(join(projects, b)).mtimeMs - statSync(join(projects, a)).mtimeMs,
			)[0];
		}
	} catch {
		// No projects dir yet, or unreadable — fall through to the formula.
	}
	return slugFormula(path);
}

// ---------------------------------------------------------------------------
// Screen-capture token — issued once at screen-capture-server startup,
// stored 0600 at ~/.config/sutando/screen-capture-token.  Callers include
// it in the X-Sutando-Capture-Token header so a browser page on loopback
// cannot reach the /capture endpoint (browsers cannot set custom headers on
// no-cors requests or read local files).
// ---------------------------------------------------------------------------

const _CAPTURE_TOKEN_PATH = join(process.env.HOME || '', '.config', 'sutando', 'screen-capture-token');

/**
 * Read the screen-capture server token from disk.  Returns the token string
 * or undefined if the file is absent (server not running or not yet started).
 * Result is NOT cached — the server may rotate the token on restart.
 */
export function readCaptureToken(): string | undefined {
	try {
		const tok = readFileSync(_CAPTURE_TOKEN_PATH, 'utf8').trim();
		return tok || undefined;
	} catch {
		return undefined;
	}
}
